x-pack/plugins/security_solution/server/endpoint/routes/resolver/utils/ancestry_query_handler.ts

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SearchResponse } from 'elasticsearch';
+import { ILegacyScopedClusterClient } from 'kibana/server';
+import {
+  parentEntityId,
+  entityId,
+  getAncestryAsArray,
+} from '../../../../../common/endpoint/models/event';
+import {
+  ResolverAncestry,
+  ResolverEvent,
+  ResolverLifecycleNode,
+} from '../../../../../common/endpoint/types';
+import { createAncestry, createLifecycle } from './node';
+import { LifecycleQuery } from '../queries/lifecycle';
+import { QueryInfo } from '../queries/multi_searcher';
+import { QueryHandler } from './fetch';
+
+/**
+ * Retrieve the ancestry portion of a resolver tree.
+ */
+export class AncestryQueryHandler implements QueryHandler<ResolverAncestry> {
+  private readonly ancestry: ResolverAncestry = createAncestry();
+  private ancestorsToFind: string[];
+  private readonly query: LifecycleQuery;
+
+  constructor(
+    private levels: number,
+    indexPattern: string,
+    legacyEndpointID: string | undefined,
+    originNode: ResolverLifecycleNode | undefined
+  ) {
+    this.ancestorsToFind = getAncestryAsArray(originNode?.lifecycle[0]).slice(0, levels);
+    this.query = new LifecycleQuery(indexPattern, legacyEndpointID);
+
+    // add the origin node to the response if it exists
+    if (originNode) {
+      this.ancestry.ancestors.push(originNode);
+      this.ancestry.nextAncestor = parentEntityId(originNode.lifecycle[0]) || null;
+    }
+  }
+
+  private toMapOfNodes(results: ResolverEvent[]) {
+    return results.reduce((nodes: Map<string, ResolverLifecycleNode>, event: ResolverEvent) => {
+      const nodeId = entityId(event);
+      let node = nodes.get(nodeId);
+      if (!node) {
+        node = createLifecycle(nodeId, []);
+      }
+
+      node.lifecycle.push(event);
+      return nodes.set(nodeId, node);
+    }, new Map());
+  }
+
+  private setNoMore() {
+    this.ancestry.nextAncestor = null;
+    this.ancestorsToFind = [];
+    this.levels = 0;
+  }
+
+  private handleResponse = (searchResp: SearchResponse<ResolverEvent>) => {
+    const results = this.query.formatResponse(searchResp);
+    if (results.length === 0) {
+      this.setNoMore();
+      return;
+    }
+
+    // bucket the start and end events together for a single node
+    const ancestryNodes = this.toMapOfNodes(results);
+
+    // the order of this array is going to be weird, it will look like this
```

Comment:
not necessary, but I feel like a mini diagram could be helpful here too

Reply:
I'll add the diagram in a follow up PR so I don't have to wait for ci again haha